komposable-architecture-test/src/main/java/com/toggl/komposable/test/store/TestStore.kt

```diff
@@ -0,0 +1,415 @@
+    internal val logger: Logger,
+    internal val reflectionHandler: ReflectionHandler = PublicPropertiesReflectionHandler(),
+    val testCoroutineScope: TestScope,
+    var timeout: Duration = 100.milliseconds,
+    var exhaustivity: Exhaustivity = Exhaustivity.Exhaustive,
+) {
+    private val store: Store<State, TestReducer.TestAction<Action>>
+    internal val reducer: TestReducer<State, Action>
+    val state: State
+        get() = reducer.state
+
+    private val effectSubscriptionChannel = Channel<Unit>(capacity = BUFFERED)
+
+    init {
+        this.reducer = TestReducer(
+            reducer(),
+            initialState(),
+            testCoroutineScope,
+            effectSubscriptionChannel,
+        )
+        this.store = createStore(
+            initialState = this.state,
+            reducer = this.reducer,
+            dispatcherProvider = dispatcherProvider,
+            storeScopeProvider = { testCoroutineScope },
+        )
+    }
+
+    private val assertionRunner: AssertionRunner<State, Action>
+        get() = exhaustivity.createAssertionRunner(this)
+
+    internal interface AssertionRunner<State : Any, Action> {
+        fun assertStateChange(
+            previousState: State,
+            currentState: State,
+            assert: ((state: State) -> State)?,
+        )
+
+        fun hasReceivedActionToHandle(match: (Action) -> Boolean): Boolean
+        fun skipNotMatchingActions(match: (Action) -> Boolean)
+        fun assertEffectsAreDone()
+        suspend fun assertActionsWereReceived()
+    }
+
+    /**
+     * Controls the level of exhaustiveness for state change and action receipt assertions in [TestStore].
+     */
+    sealed class Exhaustivity {
+        /**
+         * Represents an exhaustive level where all state changes and received actions must be asserted and
+         * all effects must be done when finishing the test.
+         */
+        data object Exhaustive : Exhaustivity()
+
+        /**
+         * Represents a non-exhaustive level where state changes and received actions can be skipped.
+         * Note that actions fired by effects must still be received or skipped before new actions can be sent.
+         * The final state of the store can be asserted using [TestStore.assert] after calling [TestStore.finish]
+         */
+        data class NonExhaustive(
+            val logIgnoredReceivedActions: Boolean = true,
+            val logIgnoredStateChanges: Boolean = true,
+            val logIgnoredEffects: Boolean = true,
+        ) : Exhaustivity()
+    }
+
+    /**
+     * Sends an action to the store and asserts the resulting state change.
+     * Don't provide an assert function if you don't expect a state change.
+     *
+     * @param action The action to be sent to the store.
+     * @param assert A lambda function for asserting the resulting state.
+     */
+    suspend fun send(action: Action, assert: ((state: State) -> State)? = null) {
```

Comment:
Should we consider passing `var timeout: Duration = 100.milliseconds` and/or `var exhaustivity: Exhaustivity = Exhaustivity.Exhaustive` to this send and receive functions?

Reply:
A user of TestStore would be expected to know and change these variables before calling the respective methods. If we use them as parameters then it would be clearer.

Reply:
Although we could do that (I assume you mean changing the exhaustivity and timeout to send/receive parameters), it would pollute the send/receive signatures. We have the properties and We've introduced the test extension exactly for that, so that we can have something like:

```kotlin
store.exhaustivity = NonExhaustive(...)
store.test {
  withExhaustivity(Exhaustive) { 
    send(...)
    receive(...)
  }
}
```

The `timeout` is also more advanced usage (or at changed/configure just once), not sure if we should be making it easier to change.